File: app/src/main/java/com/github/willjgriff/ethereumwallet/ui/navigation/NavigationController.java
package com.github.willjgriff.ethereumwallet.ui.navigation;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.bluelinelabs.conductor.Controller;
import com.bluelinelabs.conductor.RouterTransaction;
import com.bluelinelabs.conductor.changehandler.FadeChangeHandler;
import com.github.wiljgriff.ethereumwallet.ui.transactions.TransactionsController;
import com.github.willjgriff.ethereumwallet.R;
import com.github.willjgriff.ethereumwallet.di.invalidation.ComponentsInvalidator;
import com.github.willjgriff.ethereumwallet.ui.navigation.di.DaggerNavigationComponent;
import com.github.willjgriff.ethereumwallet.ui.utils.UiUtils;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Will on 29/01/2017.
 */

public class NavigationController extends Controller
	implements NavigationToolbarListener {

	private static final String STATE_SELECTED_ITEM = NavigationToolbarListener.class.getName() + ";STATE_SELECTED_ITEM";

	@BindView(R.id.controller_navigation_toolbar)
	Toolbar mToolbarTitle;
	@BindView(R.id.controller_navigation_balance)
	TextView mBalance;
	@BindView(R.id.controller_navigation_container)
	ViewGroup mControllerContainer;
	@BindView(R.id.controller_navigation_bottom_navigation)
	BottomNavigationView mBottomNavigationView;

	// TODO: See if there is a better way of doing this / find it manually each time it's needed.
	private int mCurrentlySelectedMenuItemId = 0;

	@Inject
	ComponentsInvalidator mComponentsInvalidator;

	public NavigationController() {
		DaggerNavigationComponent.builder()
			.build()
			.inject(this);
	}

	@NonNull
	@Override
	protected View onCreateView(@NonNull LayoutInflater inflater, @NonNull ViewGroup container) {
		View view = inflater.inflate(R.layout.controller_navigation, container, false);
		ButterKnife.bind(this, view);

		// This is currently necessary because if the Activity gets reclaimed by the OS we loose the
		// position on the BottomNavigationMenu. Good job BottomNav...
		if (mCurrentlySelectedMenuItemId != 0) {
			mBottomNavigationView.getMenu().findItem(mCurrentlySelectedMenuItemId).setChecked(true);
		}

		if (!getChildRouter(mControllerContainer).hasRootController()) {
			switchToController(new TransactionsController());
		}

		setupBottomNavigationView();
		return view;
	}

	private void setupBottomNavigationView() {
		final NavigationControllerFactory navigationControllerFactory = new NavigationControllerFactory();
		mBottomNavigationView.setOnNavigationItemSelectedListener(item ->
			onBottomNavItemClicked(navigationControllerFactory, item));
	}

	private boolean onBottomNavItemClicked(NavigationControllerFactory navigationControllerFactory, MenuItem item) {
		mCurrentlySelectedMenuItemId = item.getItemId();
		UiUtils.hideSoftKeyboard(getView());
		Controller controller = navigationControllerFactory.getController(item);
		if (controller != null) {
			switchToController(controller);
			return true;
		}
		return false;
	}

	private void switchToController(Controller controller) {
		mComponentsInvalidator.invalidateComponents();
		controller.setTargetController(this);
		getChildRouter(mControllerContainer)
			.setRoot(RouterTransaction.with(controller)
				.pushChangeHandler(new FadeChangeHandler())
				.popChangeHandler(new FadeChangeHandler()));
	}

	@Override
	protected void onSaveInstanceState(@NonNull Bundle outState) {
		super.onSaveInstanceState(outState);
		outState.putInt(STATE_SELECTED_ITEM, mCurrentlySelectedMenuItemId);
	}

	@Override
	protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
		super.onRestoreInstanceState(savedInstanceState);
		mCurrentlySelectedMenuItemId = savedInstanceState.getInt(STATE_SELECTED_ITEM);
	}

	@Override
	public void setToolbarTitle(CharSequence toolbarTitle) {
		// TODO: Animate the changing of this title to match the FadeChangeHandler of the Controllers.
		mToolbarTitle.setTitle(toolbarTitle);
	}

	@Override
	public void setBalance(String balance) {
		mBalance.setText(balance);
	}
}

File: app/src/main/kotlin/com/github/willjgriff/ethereumwallet/ui/navigation/NavigationController.kt
package com.github.willjgriff.ethereumwallet.ui.navigation

import android.os.Bundle
import android.support.design.widget.BottomNavigationView
import android.support.v7.widget.Toolbar
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.bluelinelabs.conductor.Controller
import com.bluelinelabs.conductor.RouterTransaction
import com.bluelinelabs.conductor.changehandler.FadeChangeHandler
import com.github.willjgriff.ethereumwallet.R
import com.github.willjgriff.ethereumwallet.ui.transactions.TransactionsController
import com.github.willjgriff.ethereumwallet.ui.utils.hideSoftKeyboard
import kotlinx.android.synthetic.main.controller_navigation.view.*

/**
 * Created by williamgriffiths on 04/05/2017.
 */
class NavigationController : Controller(), NavigationToolbarListener {

    private val STATE_SELECTED_MENU_ITEM = NavigationToolbarListener::class.simpleName + ";STATE_SELECTED_MENU_ITEM"

    private lateinit var toolbar: Toolbar
    private lateinit var balance: TextView
    private lateinit var controllerContainer: ViewGroup
    private lateinit var bottomNavigationView: BottomNavigationView

    // TODO: See if there is a better way of doing this / find it manually each time it's needed.
    private var currentlySelectedMenuItemId = 0

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup): View {
        val view = inflater.inflate(R.layout.controller_navigation, container, false)
        bindViews(view)

        // This is currently necessary because if the Activity gets reclaimed by the OS we loose the
        // position on the BottomNavigationMenu. Good job BottomNav...
        if (currentlySelectedMenuItemId != 0) {
            bottomNavigationView.menu.findItem(currentlySelectedMenuItemId).isChecked = true
        }

        if (!getChildRouter(controllerContainer).hasRootController()) {
            switchToController(TransactionsController())
        }

        setupBottomNavigationView()
        return view
    }

    private fun bindViews(view: View) {
        toolbar = view.controller_navigation_toolbar
        balance = view.controller_navigation_balance
        controllerContainer = view.controller_navigation_container
        bottomNavigationView = view.controller_navigation_bottom_navigation
    }

    private fun setupBottomNavigationView() {
        bottomNavigationView
                .setOnNavigationItemSelectedListener { item -> bottomNavItemClicked(NavigationControllerFactory(), item) }
    }

    private fun bottomNavItemClicked(navigationControllerFactory: NavigationControllerFactory, item: MenuItem): Boolean {
        currentlySelectedMenuItemId = item.itemId
        view?.hideSoftKeyboard()
        val controller = navigationControllerFactory.getController(item)
        if (controller != null) {
            switchToController(controller)
            return true
        }
        return false
    }

    private fun switchToController(controller: Controller) {
        controller.targetController = this
        getChildRouter(controllerContainer)
                .setRoot(RouterTransaction.with(controller)
                        .pushChangeHandler(FadeChangeHandler())
                        .popChangeHandler(FadeChangeHandler()))
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(STATE_SELECTED_MENU_ITEM, currentlySelectedMenuItemId)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        currentlySelectedMenuItemId = savedInstanceState.getInt(STATE_SELECTED_MENU_ITEM)
    }

    override fun setToolbarTitle(toolbarTitle: CharSequence) {
        // TODO: Animate the changing of this title to match the FadeChangeHandler of the Controllers.
        this.toolbar.title = toolbarTitle
    }

    override fun setBalance(balance: String) {
        this.balance.text = balance
    }
}